src/runtime/virtcontainers/utils/utils.go

```diff
@@ -306,6 +306,36 @@ func ConvertAddressFamily(family int32) pbTypes.IPFamily {
 	}
 }
 
+func waitProcessUsingWaitLoop(pid int, timeoutSecs uint, logger *logrus.Entry) bool {
+	secs := time.Duration(timeoutSecs)
+	timeout := time.After(secs * time.Second)
+
+	for {
+		// Check if the process is running periodically to avoid a busy loop
+		// "A watched pot never boils" and an unwaited-for process never appears to die!
+		waitedPid, err := syscall.Wait4(pid, nil, syscall.WNOHANG, nil)
+
+		if waitedPid == pid && err == nil {
+			return false
+		}
+
+		if err := syscall.Kill(pid, syscall.Signal(0)); err != nil {
+			return false
+		}
+
+		select {
+		case <-time.After(50 * time.Millisecond):
```

Comment:
In the end, what you are trying to do is to turn a `do { sleep(); } while (!dead)` into a `while (!dead) { sleep(); }`

That makes sense, but it's a win only if the process dies very quickly.

What about having a `time.After(delay)` where delay starts at, say, 1ms, and the next delay is max(delay*5, 50). So you'd have 1, 5, 25, 50 wait times, and then loop every 50ms. That way, you reap a quickly dying process very quickly (which I think was your intent) even if it takes, say, 5ms to die and no 0.25.

Reply:
This would be much better indeed but I'd like this improvement to sit in a separate commit (either preliminary or follow-up).

Reply:
Nice, added exponential backoff in a new commit.